### PR TITLE
feat: add DESCRIBE functionality for connectors

### PR DESCRIPTION
**Pull request by agavra**
### Description 
Implements functionality to describe a connector.

### Testing done 
Unit testing and local test:

```
ksql> DESCRIBE CONNECTOR "jdbc-connector2";

Name                 : jdbc-connector2
Type                 : io.confluent.connect.jdbc.JdbcSourceConnector
State                : RUNNING
WorkerId             : 10.200.7.69:8083

 Task ID | State   | Error Trace
---------------------------------
 0       | RUNNING |
---------------------------------

 Source Name           | Kafka Topic | Type
---------------------------------------------
 JDBC_CONNECTOR2_USERS | jdbc2-users | TABLE
---------------------------------------------
```

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java [hunk 1]
```diff
@@ -43,6 +43,7 @@
 import io.confluent.ksql.rest.entity.ArgumentInfo;
 import io.confluent.ksql.rest.entity.CommandStatusEntity;
 import io.confluent.ksql.rest.entity.ConnectorList;
+import io.confluent.ksql.rest.entity.ConnectorDescription;
```
Comment (cpettitt-confluent) [inline ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: checkstyle failed for this due to it not being sorted lexicographically.

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultConnectClient.java [hunk 2]
```diff
@@ -119,6 +121,30 @@ public DefaultConnectClient(final String connectUri) {
...
+      final ConnectResponse<ConnectorStateInfo> connectResponse = Request
+          .Get(connectUri.resolve(CONNECTORS + "/" + connector + STATUS))
+          .socketTimeout(DEFAULT_TIMEOUT_MS)
+          .connectTimeout(DEFAULT_TIMEOUT_MS)
+          .execute()
+          .handleResponse(
+              createHandler(HttpStatus.SC_OK, ConnectorStateInfo.class, Function.identity()));
+
+      connectResponse.error()
+          .ifPresent(error -> LOG.warn("Did not CREATE connector {}: {}", connector, error));
```
Comment (cpettitt-confluent) [inline ksql-engine/src/main/java/io/confluent/ksql/services/DefaultConnectClient.java]: Maybe "Could not query status of connector {}: {}" ?

File: ksql-engine/src/test/java/io/confluent/ksql/connect/ConnectConfigServiceTest.java [hunk 3]
```diff
@@ -101,13 +101,15 @@ public void shouldCreateConnectorFromConfig() throws InterruptedException {
...
 
   @Test
-  public void shouldWakeupConsumerBeforeShuttingDown() {
+  public void shouldWakeupConsumerBeforeShuttingDown() throws InterruptedException {
     // Given:
+    final CountDownLatch noMoreLatch = new CountDownLatch(1);
     setupConfigService();
-    givenNoMoreRecords(when(consumer.poll(any())));
+    givenNoMoreRecords(when(consumer.poll(any())), noMoreLatch);
     configService.startAsync().awaitRunning();
```
Comment (cpettitt-confluent) [inline ksql-engine/src/test/java/io/confluent/ksql/connect/ConnectConfigServiceTest.java]: Minor: Isn't this where the action happens? If so, would this be better suited for "when"?

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java [hunk 4]
```diff
@@ -643,6 +648,40 @@ private void printQueryDescription(final QueryDescription query) {
...
+              .map(task -> ImmutableList.of(
+                  String.valueOf(task.id()),
+                  task.state(),
+                  ObjectUtils.defaultIfNull(task.trace(), ""))))
+          .build();
+      taskTable.print(this);
+      writer().println();
+    }
+
+    if (!description.getSources().isEmpty()) {
```
Comment (cpettitt-confluent) [inline ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: What about SINK and UNKNOWN? Do these come in a different patch?

Also, I assume that even though we have separate CREATE SOURCE CONNECTOR xxx and CREATE SINK CONNECTOR xxx statements that the namespace for both are the same for the status query to work unambiguously?

File: ksql-engine/src/test/java/io/confluent/ksql/services/DefaultConnectClientTest.java [hunk 5]
```diff
@@ -46,6 +48,12 @@
       ImmutableList.of(new ConnectorTaskId("foo", 1)),
       ConnectorType.SOURCE
   );
+  private static final ConnectorStateInfo SAMPLE_STATUS = new ConnectorStateInfo(
+      "foo",
+      new ConnectorState("state", "worker", "msg"),
+      ImmutableList.of(),
```
Comment (cpettitt-confluent) [inline ksql-engine/src/test/java/io/confluent/ksql/services/DefaultConnectClientTest.java]: Should we test with a task to verify it is propagated successfully?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/ConnectorDescription.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+public class ConnectorDescription extends KsqlEntity {
+
+  private final String type;
+  private final ConnectorStateInfo status;
+  private final List<SourceDescription> sources;
+
+  @JsonCreator
+  public ConnectorDescription(
+      @JsonProperty("statementText")  final String statementText,
+      @JsonProperty("type")           final String type,
```
Comment (cpettitt-confluent) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/ConnectorDescription.java]: When I first saw this I thought it might be redundant with ConnectorStateInfo.type. Maybe worth a note that this is the connector class (I believe?)?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/ConnectorDescription.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    return status;
+  }
+
+  public List<SourceDescription> getSources() {
+    return sources;
+  }
+
+  @Override
+  public boolean equals(final Object o) {
+    if (this == o) {
```
Comment (cpettitt-confluent) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/ConnectorDescription.java]: It feels like type might be part of the identity for this class. Any reason to leave it out?

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/execution/DescribeConnectorExecutorTest.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+import org.apache.kafka.connect.runtime.rest.entities.ConnectorStateInfo.TaskState;
+import org.apache.kafka.connect.runtime.rest.entities.ConnectorType;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DescribeConnectorExecutorTest {
```
Comment (cpettitt-confluent) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/execution/DescribeConnectorExecutorTest.java]: Would it have been possible to just inject the ConnectClient into TestServiceContext here?

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java (hunk 1 above)
Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: Oops! Merge conflict with my other PR. Thanks for catching this :)

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java (hunk 4 above)
Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: Terminology here is a little confusing. Internally, "source" refers to anything that is a Stream or a Table, not necessarily something that produces data (e.g. a source connector). So `description.getSources()` returns everything, including sinks.

Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: > Also, I assume that even though we have separate CREATE SOURCE CONNECTOR xxx and CREATE SINK CONNECTOR xxx statements that the namespace for both are the same for the status query to work unambiguously?

Yes. Later we may want to introduce specific behavior for each of them.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/ConnectorDescription.java (hunk 6 above)
Comment (agavra) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/ConnectorDescription.java]: yes - I will rename this to `connectorClass` it might be easier to grok

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/ConnectorDescription.java (hunk 7 above)
Comment (agavra) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/ConnectorDescription.java]: There is no good reason! I added it after I generated hashcode and equals and forgot to update it 🙂

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/execution/DescribeConnectorExecutorTest.java (hunk 8 above)
Comment (agavra) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/execution/DescribeConnectorExecutorTest.java]: I'm not sure I understand the suggestion - `TestServiceContext` has more than what's necessary for this unit test, it only requires usage of `ConnectClient`, so why inject it?

File: ksql-engine/src/test/java/io/confluent/ksql/connect/ConnectConfigServiceTest.java (hunk 3 above)
Comment (agavra) [reply ksql-engine/src/test/java/io/confluent/ksql/connect/ConnectConfigServiceTest.java]: I suppose so! I thought that the "action" happens when `stopAsync()` is called (e.g. it calls `triggerShutdown` which in turn causes the wakeup/close, but I think it makes sense to put it in "when' as well. Will make that change.

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java (hunk 4 above)
Comment (cpettitt-confluent) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: Speaking from a naive user's perspective (which I am right now :)), I'm a bit confused on what's being displayed. Maybe a doc update with this patch would make it clearer?

For example, what is a task in this context? Is it the thing sucking data out or pushing data into a DB? Is it a StreamTask? Something else?

Is "source" the stream / table that contains the CDC data for sources? Is it the stream / table used to feed data into the sink database? Source might be a bit overloaded because we're using it as the syntax in the "CREATE SOURCE CONNECTOR" statement, but I think source means something different here?

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/execution/DescribeConnectorExecutorTest.java (hunk 8 above)
Comment (cpettitt-confluent) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/execution/DescribeConnectorExecutorTest.java]: I didn't read this close enough. I was hoping it would simplify set up a bit, but it doesn't look like it would. We can leave this as is.

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java (hunk 4 above)
Comment (cpettitt-confluent) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: BTW, if this is already documented somewhere or the user is expected to understand the distinctions having gone through setting this up then I'm fine closing this out. I'm coming at this having only briefly interacted with connectors.

Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: It's definitely confusing because this is the point of intersection between connect and KafkaStreams, both of which have pretty overlapping sets of terminologies.

- **sources** refer to the KSQL concept of a source, which is any Stream or Table irrespective of whether data is being written into or read from
- **type** refers to the connect concept of `ConnectorType`, which can be `SOURCE`, `SINK` or `UNKNOWN` and this refers to whether the connector is moving data into or out of kafka
- **task** refers to a a connect worker task (https://docs.confluent.io/current/connect/concepts.html#connect-tasks) - so your intuition about it being the thing that sucks data out or pushes data into is correct (it is not a StreamTask, though it is in essence actually very similar)

So I think, correct me if I'm wrong, the confusion here is that while "type" and "task" are both connect concepts in the context, "sources" is actually using the KSQL concept. Perhaps it clarifies things if the title for that portion was `KSQL Sources`.

Let me know what you think, and thanks for the detailed thoughts!

Comment (cpettitt-confluent) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: I think I'm even more confused on type now :). In your example in the test code below I see a top level type, which is the connector class and I see a type in the sources table, which appears to be STREAM or TABLE. I'm not seeing a type that indicates SOURCE, SINK, UNKNOWN.

I think you're right that the KSQL use of "source" is confusing, and that goes beyond the scope of this commit, so I don't really want to hold this up. It's probably fine to rename this KSQL Sources for now and perhaps revisit the name "source" in the context of KSQL if I'm not the only one confused :).

BTW, an example of how a source and a sink connector would be described might be helpful as a way to clear up some of my lingering confusion. If you have an example handy for both, would you mind sending them my way?

Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: Ah the example in the testing done section is out of date... I actually confused myself reading it! Here is a more recent one (and a sample of how a SINK connector would look like):
```
ksql> DESCRIBE CONNECTOR "jdbc-connector3";

Name                 : jdbc-connector3
Class                : io.confluent.connect.jdbc.JdbcSourceConnector
Type                 : source
State                : RUNNING
WorkerId             : 10.200.7.69:8083

 Task ID | State   | Error Trace
---------------------------------
 0       | RUNNING |
---------------------------------

 KSQL Source Name      | Kafka Topic | Type
---------------------------------------------
 JDBC_CONNECTOR3_USERS | jdbc2-users | TABLE
---------------------------------------------

ksql> DESCRIBE CONNECTOR "elsatic-sink";

Name                 : elastic-sink
Class                : io.confluent.connect.elasticsearch.ElasticsearchSinkConnector
Type                 : sink
State                : RUNNING
WorkerId             : 10.200.7.69:8083

 Task ID | State   | Error Trace
---------------------------------
 0       | RUNNING |
---------------------------------
```
The sink connectors are not yet implemented, but it'll look something like that (perhaps we can even show the KSQL sources that it reads from)

Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: Originally the type that indicated `SOURCE|SINK|UNKNOWN` was hidden from this view (it was present in the JSON response, but not the tabular one)
